### DISCUSSION TITLE
Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined.

### DISCUSSION THREAD

**Discussion by FBosler** [vercel/next.js#17452] [Help]
I started running Facebook ads some days ago and I noticed always when setting them up, my sentry goes haywire shortly thereafter and spits out errors like the following:

```
Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined.
```

![image](https://user-images.githubusercontent.com/14016698/94595031-8b417f80-028a-11eb-8854-6185ed7aacb1.png)
This is really f'ing up my quotas and I have no idea how to even go about debugging the errors. Apparently, it's hundreds of users that are affected, but those users don't show up in any sort of tracking.

Does anybody have an idea how to debug this? The stack trace I see on sentry is pretty mumble jumble.

```
Error: Minified React error #130; visit https://reactjs.org/docs/error-decoder.html?invariant=130&args[]=undefined&args[]= for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
arguments: [{"message":"Minified React error #130; visit https://reactjs.org/docs/error-decoder.html?invariant=130&args[]=undefined&args[]= for the full message or use the non-minified dev environment for full errors and additional helpful warnings.","name":"Error","stack":"Error: Minified React error #130; visit https://reactjs.org/docs/error-decoder.html?invariant=130&args[]=undefined&args[]= for the full message or use the non-minified dev environment for full errors and additional helpful warnings.\n    at Vu (https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:127902)\n    at d (https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:70693)\n    at h (https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:71979)\n    at https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:74067\n    at Ia (https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:83583)\n    at ko (https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:124250)\n    at xu (https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:114740)\n    at wu (https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:114665)\n    at du (https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:111667)\n    at ou (https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:108411)\n    at Yu (https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:129963)\n    at https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:131208\n    at mu (https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:112017)\n    at tc (https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:131193)\n    at Object.t.hydrate (https://www.viteach.de/_next/static/chunks/framework.2534f472621f7cfbb93e.js:1:134117)\n    at https://www.viteach.de/_next/static/runtime/main-164716d3fdd13b14c611.js:1:9127\n    at f (https://www..."}]
```

And yea, I do upload source maps to sentry. Any pointers super welcome!!

**Comment by IRediTOTO** [comment]
This error from your code, just that. Let's find variable which have value undefine.

**Reply by FBosler** [reply]
Thx, @IRediTOTO any advice on how to find that error? As I said, it never occurs during regular sessions, it only happens with the Facebook scraper.

**Comment by FBosler** [comment]
I am "seeing" these errors again (only via sentry). Would love to get some help on this, I really think its somewhere in the framework and not my code. Help will be rewarded with a good old Munich beer. :D

![image](https://user-images.githubusercontent.com/14016698/94700954-61e12c00-033c-11eb-803f-8de7ac610bc2.png)

**Comment by lfades** [comment]
@FBosler Somewhere in your components you're not returning a component, you should be looking for something like this:

```js
function Component() {
  return undefined
}

// In some other component, this will break.
<Component />
```

If you know which page has the error, go there. Then if it works as expected, try running actions on that page that may change the state of your components until it fails. It can be your own components, a third party library, or an older version of Next.js. Hard to tell.

**Reply by FBosler** [reply]
@lfades 

Thank you very much for getting back to me! What strikes me really as odd, is that I don't get to see a path or useful stacktrace for those errors:

for example, in the following image, one can see clearly that the last error has a nice path and a useful error message. That does not hold true for the other new Promise(<anonymous>) errors

![image](https://user-images.githubusercontent.com/14016698/94776988-fe4e1180-03c2-11eb-9048-e2364fd08d81.png)

*Additionally:* All the IPs causing these errors are actually all from the Facebook ISP
![image](https://user-images.githubusercontent.com/14016698/94777210-5f75e500-03c3-11eb-8e66-c9c0e5a8ce45.png)

I can not even reproduce the error from my end of things ...

**Reply by FBosler** [reply]
@lfades this is actually the landing page and there are not that many actions a user can take. I checked all the components on the page. None of them can return `undefined`. Additionally, the hit does not show up in tracking. I.e. tag manager and such is never started.

I would simply ignore it, but the problem is that the extreme number of errors blows my sentry quotas.

**Reply by FBosler** [reply]
@lfades 
```
      ctx: { err, pathname: page, query, asPath, AppTree },
    }
    return Promise.resolve(
      props.props ? props.props : loadGetInitialProps(App, appCtx)
    ).then((initProps) =>
      doRender({
        ...props,
        err,
        Component: ErrorComponent,
        props: initProps,
      })
```

https://github.com/vercel/next.js/blob/49a59b1ad2fde90ddcba01d2f7eb0c779ab5d9ff/packages/next/client/index.tsx#L401

is where the error is thrown (at doRender). Does this seem to indicate that it's got something to do with the ErrorComponent? And any idea how to further debug this?

Will send you a Munich Beer 🍻  via mail if we get this resolved :D

Cheers
Fabian

**Comment by lfades** [comment]
If the error is so common you **should** be able to reproduce it on production, Idk if you're using Vercel but if that's the case then when you visit the page with the error it should appear on logs (in the project logs if it's a server error), or in the browser console if it's browser only.

We haven't seen any errors like this being reported for Next.js so this is most likely something in your app. If you don't see any logs then consider sending more metadata to the sentry error that may help you understand what's happening, add some console logs that you can see from your production logs and figure out what's happening.

**Reply by FBosler** [reply]
@lfades Thx for the reply, yea unfortunately I really am not able to reproduce it manually. When it happens, it only happens for:

- Facebook ISP (i.e. all the IPs originate from Facebook)
- Does not fire any tag manager tags
- Happens typically after I create a Facebook Ad
=> Makes me think that those are some kind of automated tests exhibiting a special behavior.

I am totally open about sending more metadata to sentry, but it feels like I would have to do that somewhere in the framework, no?

Cheers
Fabian

**Reply by lfades** [reply]
Nope, if you're using Sentry then you should be able to have total control over the data and errors you send to it, right?

**Reply by FBosler** [reply]
@lfades I do agree in theory (and have done so in Python-based projects). However, here I am not so sure. I am using [`@sentry/node` ](https://www.npmjs.com/package/@sentry/node)

this is how I initialize sentry in `_app.js`
```
Sentry.init({
    enabled: process.env.NODE_ENV === 'production',
    dsn: process.env.SENTRY_DSN,
});

export default function App({ Component, pageProps, err }) {
...
```

I then add the userId in my `UserProvider` like this:
```
sentry.configureScope(function (scope) {
  scope.setUser({ id: data._id });
});
```

I am not actually actively sending anything manually.

**Reply by lfades** [reply]
I haven't really used Sentry too much, from what I understand they capture all browser errors, so in this case the error may be unrelated to your app (and you may want to ignore it), or at least change it in a way you can handle the error before it gets to Sentry.

**Reply by FBosler** [reply]
@lfades Yea, captures all the errors, there is also some degree of customization possible (and additionally, you can send errors yourself). My main problem here is to handle the error I would have to understand where and especially how it is originating, which, unfortunately, I don't.
I find it super strange, as there is no useable stack trace or breadcrumbs for this particular kind of error (which I get for all other application-level caused errors)

If you can think of anything specific I can do on my end, let me know.

**Reply by lfades** [reply]
Reproduce it yourself in production, open the devtools console, go to the place where the error is probably happening, and once you find it you would have something to work with.

**Reply by FBosler** [reply]
@lfades I wish I could. Unfortunately, I can't reproduce the error as it only seems to appear for some crawling devices Facebook is using. If I could reproduce it, I could solve it easily.

**Reply by lfades** [reply]
Then ignore it, Sentry most likely provides a way of doing that 👍

**Reply by FBosler** [reply]
@lfades Again, I wish :) You can ignore it on the receiving end, but this is more a kind of muting, which still counts against one's quota. Doing it on the application level requires (again) knowing where the error occurs.

**Reply by lfades** [reply]
@FBosler If you switch `_app.js` to be a class, you can add `componentDidCatch` to it, and it should catch most of the errors in your app. Hopefully the Sentry client may also have a `onError` or something like that to filter the errors before sending them.

**Reply by FBosler** [reply]
@lfades Ahhhhh that is nice! I think that might help, let me try that. Thx!
